hi-ml/src/health_ml/training_runner.py

```diff
@@ -0,0 +1,238 @@
+
+    def run_regression_test(self) -> None:
+        if self.container.regression_test_folder:
+            with logging_section("Regression Test"):
+                # Comparison with stored results for cross-validation runs only operates on child run 0. This run
+                # has usually already downloaded the results for the other runs, and uploaded files to the parent
+                # run context.
+                regression_metrics_str = self.container.regression_metrics
+                regression_metrics = regression_metrics_str.split(',') if regression_metrics_str else None
+                # TODO: user should be able to override this value
+                crossval_arg_name = self.container.CROSSVAL_INDEX_ARG_NAME
+
+                logging.info("Comparing the current results against stored results.")
+                if self.is_crossval_disabled_or_child_0():
+                    if is_running_in_azure_ml():
+                        if PARENT_RUN_CONTEXT is not None:
+                            metrics = get_metrics_for_hyperdrive_run(
+                                child_run_arg_name=crossval_arg_name,
+                                run=PARENT_RUN_CONTEXT,
+                                keep_metrics=regression_metrics,
+                            )
+                        else:
+                            metrics = get_metrics_for_run(run=RUN_CONTEXT, keep_metrics=regression_metrics)
+
+                        if metrics:
+                            metrics_filename = self.container.outputs_folder / REGRESSION_TEST_METRICS_FILENAME
+                            logging.info(f"Saving metrics to {metrics_filename}")
+                            metrics_filename.write_text(json.dumps(metrics))
+
+                    compare_folders_and_run_outputs(
+                        expected=self.container.regression_test_folder,
+                        actual=self.container.outputs_folder,
+                        csv_relative_tolerance=self.container.regression_test_csv_tolerance,
+                    )
+                else:
+                    logging.info("Skipping as this is not cross-validation child run 0.")
+        else:
+            logging.info("Skipping regression test, no available results to compare with.")
+
+    def run(self) -> None:
+        """
+        Driver function to run a ML experiment
+        """
+        self.init_training()
+
+        if not self.container.run_inference_only:
+            # Backup the environment variables in case we need to run a second training in the unit tests.
+            environ_before_training = dict(os.environ)
+
+            with logging_section("Model training"):
+                self.run_training()
+
+            self.end_training(environ_before_training)
+
+        self.init_inference()
+
+        with logging_section("Model validation"):
+            self.run_validation()
+
+        with logging_section("Model inference"):
+            self.run_inference()
```

Comment:
Do we still want to call `init_inference` and `run_inference` in TrainingRunner?

Reply:
yes, we want to keep the legacy behaviour: When training a model, we will always also call inference on the test set.